src/traits.rs

```diff
@@ -225,3 +228,43 @@ pub trait Encoding: Sized {
     /// Encode to little endian bytes.
     fn to_le_bytes(&self) -> Self::Repr;
 }
+
+/// Support for optimized squaring
+pub trait Square: Sized
+where
+    for<'a> &'a Self: core::ops::Mul<&'a Self, Output = Self>,
+{
+    /// Computes the same as `self.mul(self)`, but may be more efficient.
+    fn square(&self) -> Self {
+        self * self
+    }
+}
+
+/// Constant-time exponentiation.
+pub trait Pow<Exponent> {
+    /// Raises to the `exponent` power.
+    fn pow(&self, exponent: &Exponent) -> Self;
+}
+
+impl<T: PowBoundedExp<Exponent>, Exponent: Bounded> Pow<Exponent> for T {
+    fn pow(&self, exponent: &Exponent) -> Self {
+        self.pow_bounded_exp(exponent, Exponent::BITS)
+    }
+}
+
+/// Constant-time exponentiation with exponent of a bounded bit size.
+pub trait PowBoundedExp<Exponent> {
+    /// Raises to the `exponent` power,
+    /// with `exponent_bits` representing the number of (least significant) bits
+    /// to take into account for the exponent.
+    ///
+    /// NOTE: `exponent_bits` may be leaked in the time pattern.
+    fn pow_bounded_exp(&self, exponent: &Exponent, exponent_bits: usize) -> Self;
+}
+
+/// Constant-time inversion.
+pub trait Inv: Sized {
+    /// Computes the inverse. Returns [`CtOption`] that evaluates to `None`
+    /// if the number is not invertible.
+    fn inv(&self) -> CtOption<Self>;
+}
```

Comment:
You might look to the trait design in `elliptic-curve` for this:

https://docs.rs/elliptic-curve/0.12.3/elliptic_curve/ops/trait.Invert.html

Notably that design has an associated `Output` type which avoids the `CtOption` when the value being inverted is non-zero:

https://docs.rs/elliptic-curve/0.12.3/src/elliptic_curve/scalar/nonzero.rs.html#169-181

Reply:
So, are you suggesting to rename `Inv` to `Invert` and add the `Output` associated type? I can do that.

Reply:
I'm fine with either name but I think it's nice to have the associated `Output` to support infallible inversions

Reply:
Done. Can't implement for `NonZero<DynResidue>` though since it does not implement `Zero`. 

Also, as a side note: would it make sense to have a `pub(crate) NonZero::new_unchecked()`, to skip the `unwrap()` in the cases like `impl Invert for NonZero<Residue>`? Or is having `unwrap()` easier to audit?

Reply:
> Can't implement for NonZero<DynResidue> though since it does not implement Zero.

Seems like it should be able to?

> Also, as a side note: would it make sense to have a pub(crate) NonZero::new_unchecked(),

Sure

Reply:
`Zero` needs a `ZERO` constant, and we can't make that for `DynResidue` since it can have different parameters

Reply:
Aah yeah, it would need valtrees-capable const generics at the very least, to allow passing in `DynResidueParams` at compile time. So no go for now anyway.

Reply:
> Also, as a side note: would it make sense to have a pub(crate) NonZero::new_unchecked(),

Just for certainty, do you want me to add it in this PR?

Reply:
Yeah that's fine

Reply:
It's fine that I add it or it's fine that I don't? :) I don't think it's worth doing in this PR

Reply:
Yeah sure, either way